ISPN-8962 PreferAvailabilityStrategy: Rely less on the stable topology

danberindei opened the pull request:
https://issues.jboss.org/browse/ISPN-8962
https://issues.jboss.org/browse/ISPN-8993

danberindei commented:
Updated to recognize more situations when multiple topologies represent the same partition and rebased on top of Ryan's ISPN-8903 changes.

danberindei commented:
Added a separate commit for https://issues.jboss.org/browse/ISPN-8993

tristantarrant commented:
I assume https://ci.infinispan.org/job/Infinispan/job/PR-5866/5/testReport/junit/org.infinispan.server.hotrod/HotRodMergeTest/testNewTopologySentAfterOverlappingMerge_DIST_SYNC_/ is unrelated ?

danberindei commented:
I forgot about that test, it's definitely related.

danberindei commented:
@tristantarrant `HotRodMergeTest` was failing because I had changed the way the preferred topology is selected, and the new pick required an extra rebalance. I pushed some more changes to avoid the extra rebalance if possible. I also changed `PreferConsistencyStrategy` to make it use the read CH for conflict resolution (for rebalance phases `READ_NEW` or `READ_ALL`).

I was also getting a different error when running locally and I tracked it to the other caches defined by the HR server that also have conflict resolution enabled (___protobuf_metadata, org.infinispan.CONFIG, __global_tx_table__, org.infinispan.COUNTER), because 2 test methods use the same servers and the 2nd test wasn't waiting for all those caches to finish conflict resolution. I've created https://issues.jboss.org/browse/ISPN-9009 to test this scenario properly.

tristantarrant commented:
is https://ci.infinispan.org/job/Infinispan/job/PR-5866/8/testReport/junit/org.infinispan.functional/FunctionalDistributionTest/testDistributionFromPrimaryOwner/ related ?

danberindei commented:
@tristantarrant almost certainly not related, https://issues.jboss.org/browse/ISPN-7098

tristantarrant commented on core/src/main/java/org/infinispan/topology/ClusterTopologyManagerImpl.java:
```diff
@@ -502,7 +502,7 @@ private void recoverClusterStatus(int newViewId, final List<Address> clusterMemb
       LimitedExecutor cs = new LimitedExecutor("Merge-" + newViewId, stateTransferExecutor, maxThreads);
       for (final Map.Entry<String, Map<Address, CacheStatusResponse>> e : responsesByCache.entrySet()) {
-         CacheJoinInfo joinInfo = e.getValue().values().stream().findAny().get().getCacheJoinInfo();
+         CacheJoinInfo joinInfo = e.getValue().values().iterator().next().getCacheJoinInfo();
```
I wish ordered collections had a first() method that encapsulates iterator().next()

danberindei replied:
I don't think it's that bad, `first()` would probably return an `Optional` like the one in `Stream` so you'd need an extra method call anyway. The only advantage would be replacing the `NoSuchElementException` with our own, but then that would be more verbose, and we assume at least one response anyway.

tristantarrant commented:
Looks good to me. I actually understood it :smile:

tristantarrant commented on core/src/main/java/org/infinispan/topology/ClusterCacheStatus.java:
```diff
@@ -792,11 +793,17 @@ public synchronized void startQueuedRebalance() {
          updateTopology = cacheTopology.getPendingCH() != null;
+         // After a cluster view change that leaves only 1 node, we don't need either a topology update or a rebalance
+         // but we must still update the stable topology
+         updateStableTopology =
```
Wondering if it makes sense to have methods on CacheTopology that encapsulate some of the logic (I haven't checked about repeated uses).
e.g cacheTopology.needsStableTopologyUpdate(stableTopology)

danberindei replied:
I'd like to avoid adding methods to `CacheTopology` that the application shouldn't care about, because it's exposed in `DistributionManager` and that makes changes harder.

For 9.3 I'd actually like to move internals like `ClusterCacheStatus` and `ClusterTopologyManagerImpl` away from using `CacheTopology`, so I can deprecate currentCH/pendingCH/unionCH/persistentUUIDs. The internal-only replacement would be a good location for such methods, though.

wburns commented on core/src/test/java/org/infinispan/conflict/impl/BaseMergePolicyTest.java:
```diff
@@ -144,45 +143,24 @@ public void testPartitionMergePolicy() {
-      return maxTopologySender;
+      Map<Address, CacheStatusResponse> statusResponses =
+         Arrays.stream(caches).map(c -> new KeyValuePair<>(address(c), getCacheStatus(c)))
+               .collect(Collectors.toMap(KeyValuePair::getKey, KeyValuePair::getValue));
```
Can be just: `Arrays.stream(caches).collect(Collectors.toMap(this::address, this::getCacheStatus));`

danberindei replied:
Here too, I needed a `LinkedHashMap` at some point but it's too much now

wburns commented on core/src/test/java/org/infinispan/partitionhandling/impl/PreferAvailabilityStrategyTest.java:
```diff
@@ -53,162 +73,199 @@
    @BeforeMethod(alwaysRun = true)
    public void setup() {
+      mockitoSession = Mockito.mockitoSession()
+                              .strictness(Strictness.STRICT_STUBS)
```
The default is `STRICT_STUBS` in 2.7. Can you clarify what this is doing for us?

danberindei replied:
I thought it was only going to be the default in 3.0, I'll remove it

danberindei replied:
The explicit `MockitoSession` seems to help by allowing me to skip `verify()` for method calls already expected with `when()`. I don't know why, as I thought that was a feature of STRICT_SUBS, but I'll keep it for now.

wburns replied:
Oh because you are using `noMoreInteracations`. Yeah I guess the Mock must use WARN by default in this case, but the Session default is STRICT_SUBS - kinda odd.

wburns commented on core/src/test/java/org/infinispan/topology/ClusterCacheStatusTest.java:
```diff
@@ -0,0 +1,147 @@
+
+   @BeforeMethod(alwaysRun = true)
+   public void setup() {
+      mockitoSession = Mockito.mockitoSession().initMocks(this).strictness(Strictness.STRICT_STUBS).startMocking();
```
Same thing with this one. Also I don't see any fields annotated with `@Mock` for `initMocks`?

wburns commented on core/src/test/java/org/infinispan/topology/TestClusterCacheStatus.java:
```diff
@@ -0,0 +1,245 @@
+    */
+   public ConsistentHash readConsistentHash() {
+      switch (topology.getPhase()) {
+         case NO_REBALANCE:
```
replace this with `AvailabilityStrategy.readConsistentHash`?

wburns commented on core/src/main/java/org/infinispan/partitionhandling/impl/PreferAvailabilityStrategy.java:
```diff
@@ -93,166 +68,283 @@ public void onClusterViewChange(AvailabilityStrategyContext context, List<Addres
+               }
+            }
+            if (fold) {
+               referencePartition.senders.addAll(p.senders);
```
I kinda wonder if this would be simpler if the outer i loop was doing `for (int i = 0; i < partitions.size(); ++i)` and we removed the `referencePartition` instead?

danberindei replied:
Done

danberindei commented on core/src/main/java/org/infinispan/partitionhandling/impl/PreferAvailabilityStrategy.java:
```diff
@@ -93,166 +68,283 @@ public void onClusterViewChange(AvailabilityStrategyContext context, List<Addres
+      }
+
+      // Sort partitions in reverse topology id order to simplify the algorithm
+      partitions.sort((p1, p2) -> {
```
Yeah, I had a much complicated comparator but I realized it didn't matter.

danberindei commented on core/src/main/java/org/infinispan/partitionhandling/impl/PreferAvailabilityStrategy.java:
```diff
@@ -93,166 +68,283 @@ public void onClusterViewChange(AvailabilityStrategyContext context, List<Addres
-                                             CacheTopology maxTopology, Set<Address> possibleOwners) {
-      if (!context.resolveConflictsOnMerge() || possibleOwners.isEmpty() || possibleOwners.size() == 1)
-         return false;
+   private Partition selectPreferredPartition(List<Partition> partitions) {
```
Done

wburns commented:
Integrated into master, thanks @danberindei !
